### PR TITLE
ROS2 migration.

### PR DESCRIPTION
**Pull request by clalancette**
This is an omnibus port of urdf to ROS2.  The original code
for this came from the https://github.com/ros2/robot_model
repository, from the following commits:

7c17f1e0e20b58de71d2c24a85c0c29128d25e45
bb68744f035873101f4d97abff473d3fc9471a5e
ae9e7dbbe7f64df30c29993956281e9ad96ae905
c9b85bec88e4327561cd4d7ac7ad85373223c97f
6bd94a8d9d26bee289b43949278075486a0e23d5
a2d3b1290dff50e516a510a2e1e463f987462e79
cdcf138ad41f2f99ccf781de8999699f8ad94d45
762cad7eb2b05c9982665f2c284cccdd77703971
4b58a823a92b5173caa1284017c3a821a665c1e0
fbd2f1f236af2ba593ee390c8b84127926e106eb
7a194ed90f6c1df66155a07e450224d6a07f1802
558ecf2412be6f2f3d4b25f435748c5a3771a9a8
d6be80354c73cb0c2a68bb63f478317963ed9a15
c177e3921e592d2b31d0e7edc23b3b4748a53e1a
e6c82d8b89d8967b28d7bc2ece9d5af948e834a4
3cbb34add4c9dbc2e2b9a7c08d306f4d84601981

connects to ros2/ros2#419

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

### CODE REVIEW

File: urdf/CMakeLists.txt
Comment (mikaelarguedas) [inline urdf/CMakeLists.txt]: Nit: use c++14 like the rest of the stack

File: urdf/CMakeLists.txt
Comment (mikaelarguedas) [inline urdf/CMakeLists.txt]: Nit: leaving the parenthesis on the following line allow to reduce the diff when modifying this list

File: urdf/CMakeLists.txt
Comment (mikaelarguedas) [inline urdf/CMakeLists.txt]: use `ament_cmake_ros` to be able to respect `BUILD_SHARED_LIBS` (https://github.com/ros2/ros2/issues/306)

File: urdf/CMakeLists.txt
Comment (mikaelarguedas) [inline urdf/CMakeLists.txt:66]: use ament lint auto rather than calling out each linter? (maybe that prevents from passing the LANGUAGE to cppcheck?)

**Review by mikaelarguedas — COMMENTED** [review]
The vast majority of these changes seem to be style related. Any chance these can make it upstream to keep the diff between the ROS1 and ROS2 version minimal?

Do you have CI jobs with these changes?

Note: I didn't review the code, just looked at the CMake

**Comment by mikaelarguedas** [comment]
Just saw: https://github.com/ros2/ros2/pull/419#issuecomment-347647757 that answers my question about running CI ^^

**Comment by clalancette** [comment]
> The vast majority of these changes seem to be style related. Any chance these can make it upstream to keep the diff between the ROS1 and ROS2 version minimal?

Yeah, some of these we can definitely push upstream.  I'll take a look at doing that, though I don't necessarily think it should hold this up.

> Do you have CI jobs with these changes?

Yeah, going on here: https://github.com/ros2/ros2/pull/419#issuecomment-347647757

File: urdf/CMakeLists.txt
Comment (clalancette) [reply urdf/CMakeLists.txt]: Will fix that.

File: urdf/CMakeLists.txt
Comment (clalancette) [reply urdf/CMakeLists.txt]: Sure, will fix.

File: urdf/CMakeLists.txt
Comment (clalancette) [reply urdf/CMakeLists.txt:66]: I'll give it a try.  It may involve more style changes.

File: urdf/CMakeLists.txt
Comment (clalancette) [reply urdf/CMakeLists.txt]: All right, will take a look.

File: urdf/CMakeLists.txt
Comment (mikaelarguedas) [reply urdf/CMakeLists.txt:66]: If it prevents you from passing the option you want to cppcheck, let's not bother about it because that'll result in C style changes for C++ code.

**Comment by mikaelarguedas** [comment]
> I don't necessarily think it should hold this up.

Agreed, now that the work of reapplying them here is done there's no point in holding it.

File: urdf/CMakeLists.txt
Comment (clalancette) [reply urdf/CMakeLists.txt:66]: Ah, I see.  So `ament_lint_auto` is currently passing for everything except for copyright notices.  Since we don't have https://github.com/ament/ament_lint/issues/75 fixed yet, it is getting tripped up on some small differences in how the BSD license in some of the files is laid out.  However, I can disable *just* the copyright tests, and things seem to pass, so I'll go with that for now.

Comment (mikaelarguedas) [reply urdf/CMakeLists.txt:66]: cool:

> I can disable just the copyright tests, and things seem to pass, so I'll go with that for now.

you can do this by using:
```
find_package(ament_lint_auto REQUIRED)  
set(ament_cmake_copyright_FOUND TRUE)
ament_lint_auto_find_test_dependencies()
```

File: urdf/CMakeLists.txt
Comment (mikaelarguedas) [inline urdf/CMakeLists.txt]: :+1:

**Review by mikaelarguedas — DISMISSED** [review]
lgtm with green CI (and assuming that linters pass locally with last changes)
thanks for iterating,

**Comment by mikaelarguedas** [comment]
I don't know how much work this would be, but it would be great to have the commits mentionned in the description as part of the history. It will give use better trace ability as well as ease submission of patches upstream rather than a single commit including all the changes ROS2 related or not

**Comment by clalancette** [comment]
CI was yellow on Windows with the switch to `ament_lint_auto`.  Therefore, I reverted that piece and went back to calling the tests individually for now in fcfbad6

**Comment by clalancette** [comment]
> I don't know how much work this would be, but it would be great to have the commits mentionned in the description as part of the history. It will give use better trace ability as well as ease submission of patches upstream rather than a single commit including all the changes ROS2 related or not

(sorry, I missed this comment before)

It's certainly possible, but it will be a lot more work to get it done.  I'll have to go back into the patches one-by-one, apply them, fix conflicts, then commit the results.  Potentially I should have done this to start with, but there honestly doesn't seem to be a huge amount of benefit; most of the commits are one-line commit messages, and they are really all going for a single goal of porting to ROS2.  One potential compromise would be to add the commit messages from each of those earlier commits into the omnibus commit message, which would extract most of the value.  What do you think about that?

**Comment by mikaelarguedas** [comment]
My concern is that once robot_model fork is deleted, the commits referenced in the first comments will not be introspectable.

A compromise could be to just split this PR into 3 different commits:
- style fixes
- windows compatibility
- ROS2 specific changes

This shouldn't be too much work and would at least give us proper commit history and make it easy to review and backport. What do you think about that ?

Then if we run out of time, I guess that's fine to merge as is. We will have to produce the work later on to reconciliate the repos.

**Comment by clalancette** [comment]
> A compromise could be to just split this PR into 3 different commits:
> 
>     style fixes
>     windows compatibility
>     ROS2 specific changes
> 
> This shouldn't be too much work and would at least give us proper commit history and make it easy to review and backport. What do you think about that ?

Yeah, that is a good idea, and should be pretty easy to do.  I'll make that change here and in the kdl_parser repository.

**Comment by mikaelarguedas** [comment]
> Yeah, that is a good idea, and should be pretty easy to do. I'll make that change here and in the kdl_parser repository.

:+1:

**Comment by clalancette** [comment]
OK, rebased to split it up as we discussed.  CI is in ros2/ros2#419.  Ready for another review.

File: urdf/src/model.cpp
Comment (mikaelarguedas) [inline urdf/src/model.cpp]: ~~nit: include order~~

Edit: resolved

File: urdf/test/test_robot_model_parser.cpp
Comment (mikaelarguedas) [inline urdf/test/test_robot_model_parser.cpp]: this is C++ 11 and cannot be used upstream before maybe melodic. So it should be moved to another commit

File: urdf/include/urdf/visibility_control.hpp
Comment (mikaelarguedas) [inline urdf/include/urdf/visibility_control.hpp]: This should use the same licence as the rest of this package (BSD as listed in package.xml)

File: urdf/CMakeLists.txt
Comment (mikaelarguedas) [inline urdf/CMakeLists.txt]: This should be 14 like the line below

File: urdf/package.xml
Comment (mikaelarguedas) [inline urdf/package.xml:45]: shouldn't this be ament_cmake_ros ?

File: urdf/src/model.cpp
Comment (mikaelarguedas) [inline urdf/src/model.cpp:103]: ~~style change should go in first commit~~

Edit: resolved

**Review by mikaelarguedas — COMMENTED** [review]
a few thing to move around commit-wise, otherwise only the build type and the licence need to be adressed

File: urdf/package.xml
Comment (clalancette) [reply urdf/package.xml:45]: When I try that, ament complains with:

```
RuntimeError: This script does not support the build type 'ament_cmake_ros'
Only the following build types are supported: ament_cmake, bazel, cmake, ament_python
```

**Comment by clalancette** [comment]
Other than the open question about `ament_cmake_ros`, the latest changes should fix all of the nits.

File: urdf/include/urdf/model.h
Comment (Karsten1987) [inline urdf/include/urdf/model.h]: if being ROS2 coherent, then we should have a new line between return type and function. Don't feel strong about it though.

File: urdf/package.xml
Comment (mikaelarguedas) [reply urdf/package.xml:45]: Yeah your totally right, the build type is ament_cmake. Please discard previous comment

File: urdf/include/urdf/model.h
Comment (wjwwood) [reply urdf/include/urdf/model.h]: As someone who does that a lot, I'd say it's optional. My rule is that I will separate the return type and the method name with a new line if:

- the class already does this for another function
- the function is too long for 100 characters, I'll do this first
- a function was changed to this style, I will change the others to match (consistent within a class or file)

If the functions/methods are all short (under 100 characters), then not wrapping is fine.

----

Also, in this particular case the code is from another existing project and therefore I'd say it doesn't _have_ to change, especially since this is something the linters will not catch.

By the same conventions (if you were to strictly follow the ROS 2 style) the function would be `init_xml` too, rather than `initXml`.

Comment (clalancette) [reply urdf/include/urdf/model.h]: I'd like to keep the differences with upstream to a minimum, particularly if the linters don't care about it.  So I'm going to leave it for now.

File: urdf/src/model.cpp
Comment (mikaelarguedas) [inline urdf/src/model.cpp]: I think the whitespace after the condition (between the 2 closing prenthesis) should be removed too

File: urdf/test/test_robot_model_parser.cpp
Comment (mikaelarguedas) [reply urdf/test/test_robot_model_parser.cpp]: this still needs to be addressed

File: urdf/CMakeLists.txt
Comment (mikaelarguedas) [inline urdf/CMakeLists.txt:41]: this should be moved to the windows compatibility commit

**Comment by mikaelarguedas** [comment]
ok looks like this one is almost ready too, only a couple of pending comments and we're good to go

File: urdf/test/test_robot_model_parser.cpp
Comment (clalancette) [reply urdf/test/test_robot_model_parser.cpp]: Oops, missed it.  Fixed now.

File: urdf/CMakeLists.txt
Comment (clalancette) [reply urdf/CMakeLists.txt:41]: Fixed now.

**Review by mikaelarguedas — APPROVED** [review]
lgtm, thanks for the multiple iterations

**Comment by clalancette** [comment]
Thanks, merging.

**Pull request merged**
